nemo/collections/nlp/models/language_modeling/megatron_base_model.py

```diff
@@ -322,9 +323,33 @@ def _reconfigure_val_batches(self):
         """
         Reconfigure trainer.limit_val_batches for pretraining
         """
+        # Override limit_val_batches to be a multiple of num microbatches and so there are limit_val_batches//num_micro_batches num of global batches
         if isinstance(self.trainer.limit_val_batches, int):
-            # Override limit_val_batches to be a multiple of num microbatches and so there are limit_val_batches//num_micro_batches num of global batches
             self.trainer.limit_val_batches *= get_num_microbatches()
+        else:
+            assert isinstance(self.trainer.limit_val_batches, float)
+            # Don't reconfigure if limit_val_batches is 0.0
+            if self.trainer.limit_val_batches == 0.0:
+                return
+            if self._validation_ds is not None and len(self._validation_dl) != float("inf"):
+                # len(self._validation_dl) returns len as num of microbatches
+                limit_val_micro_batches = int(len(self._validation_dl) * self.trainer.limit_val_batches)
+                if limit_val_micro_batches == 0 and self.trainer.limit_val_batches > 0.0:
+                    min_percentage = 1.0 / len(self._validation_dl)
+                    raise MisconfigurationException(
+                        f"You requested to check {self.trainer.limit_val_batches} of the val_dataloader but"
+                        f" {self.trainer.limit_val_batches} * {len(self._validation_dl)} < 1. Please increase the"
+                        f" `limit_val_batches` argument. Try at least"
+                        f" `limit_val_batches={min_percentage}`"
+                    )
+                # Make sure trainer.limit_val_batches is a multiple of num of microbatches
+                if limit_val_micro_batches < get_num_microbatches():
+                    self.trainer.limit_val_batches = get_num_microbatches()
```

Comment:
PTL handles float `limit_val_batches` in the same as we have done above (see: [data_connector.pyL449](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/trainer/connectors/data_connector.py#L449-L450) and [data_connector.py#L457](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/trainer/connectors/data_connector.py#L457-L464). 

The only difference being, after the step of casting the float `limit_val_batches` to an equivalent int is done, we are ensuring that `limit_val_batches` is a multiple of num_of_microbatches. 

This is required since in each global training_step say we extract x microbatches, and if `limit_val_batches` is not equal to x at the least or not equal to multiple of x, then this [line](https://github.com/NVIDIA/NeMo/blob/v1.22.0/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L831)  raises `StopIteration`(cause PTL bounds val_dataloder by limit_val_batches and raises StopIteration when num of microbatches extracted reaches `limit_val_batches`) in between a global step leading to a hang in case of PP, if PP 1 hit a `StopIteration` and PP2 is waiting for the output of PP1. I hope this made sense. Let me know if it's unclear.

Reply:
To answer your question on `if` and `else` condition: In the `if` case we are making sure if limit_val_batches > 0 and < num_micro_batches then it's at least equal to the num_micro_batches so that we can run 1 `validation_step` successfully. The else is for cases when limit_val_batches > num_micro_batches, almost like we are cutting out the incomplete batch that does not have as many microbatches as the `get_num_microbatches()`

Reply:
Thanks! Now it's way clearer. 

Regarding if/else: Is there an assumption that number of batches in the validation split is always greater than `get_num_micro_batches()`? Is this what allows rounding up in the `if` case?
So there is no way to pad dataset to desired length with samples that won't be counted to the final loss?

I don't think we are allowed to validate on a strict subset of the validation data. @ShriyaPalsamudram can you confirm that?

Reply:
No, we should not cut out the last incomplete batch by default. There is this flag [validation_drop_last](https://github.com/NVIDIA/NeMo/blob/df5a39583e31fd99bda371759c7954f928195594/examples/nlp/language_modeling/conf/megatron_gpt_config.yaml#L245) which decides if the last incomplete batch is dropped or not. [This section](https://github.com/NVIDIA/NeMo/blob/df5a39583e31fd99bda371759c7954f928195594/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L561-L579) handles validation_drop_last=False case in loss computation as well.

Reply:
There's no explicit condition/assumption existing in NeMo _currently_, that checks num of micro batches in the validation dataloader is always greater than `get_num_micro_batches()`. Rounding up is added in the `if` condition, cause otherwise we wont even be able to run one global step of validation for reasons mentioned above and it can lead to a hang.

In the `else` condition, if `limit_val_batches=1.0`, then there are no partial batches at all. There would be partial batches only when `limit_val_batches` is a decimal `< 1.0`. In that case what we are doing in the `else` condition is not technically chopping the batch. What I mean is, we are basically allowing to run certain percentage of the val_dataset via float `limit_val_batches`. In order to make this happen we have to translate it to equivalent int value (which is the way even PTL handles. For example if `limit_val_batches` is `0.25` and num of batches is `6` then `0.25*6=1.5` and `int(1.5)=1`, but this does not mean we are chopping that `0.5` of the second batch right ?). The logic we have is just ensuring the correct value of `limit_val_batches`. Basically in each validation step you run `limit_val_batches` num of batches and if we skipped any batch in that final value of `limit_val_batches` with which we run, then that would be like cutting out a batch.

Reply:
I agree that the dataloader shouldn't return partial batches and probably it would be better to round the number down. But I don't fully get what you mean by:
> Basically in each validation step you run limit_val_batches num of batches and if we skipped any batch in that final value of limit_val_batches with which we run, then that would be like cutting out a batch. Let me know your thoughts.